### DISCUSSION TITLE
Dynamic number of queries

### DISCUSSION THREAD

**Discussion by agriffis** [TanStack/query#493] [General] [answered]
I need to make a dynamic number of queries depending on user config:

```
config = {ids: [1, 2, 3, 4]}
results = config.ids.map(id => useQuery(id, fetch))
```

Now of course we can't use hooks this way. What we need instead is a `useQueries` hook that internally manages the list or object of ongoing queries.

For example, I wrote a hook called [usePromises](https://gist.github.com/agriffis/b0b48923130bd30a2ec4e6bddeb2e629), but it doesn't have all the stale-while-revalidate and other functionality of react-query.

Any thought of adding this to react-query, or other suggestions?

**Comment by denisborovikov** [comment, marked as answer]
Use only one useQuery and retrive the all required data in your fetch function.

```
results = useQuery(['todos', config], () => Promise.all(...))
```

**Reply by agriffis** [reply]
Thanks! But then the component won't rerender as API calls return, instead it has to wait for the entire batch, right?

**Reply by tannerlinsley** [reply]
Correct. But in reality, the whole point of parallel queries is that they realistically all need to resolve before you can do anything with them, otherwise, you could just put them in separate components.

**Reply by agriffis** [reply]
Between these two answers, I might have what I need. Thanks!

**Reply by lukaszmakuch** [reply]
Hi all!

I'm also interested in dynamically issuing multiple queries.

The reason why I'd love to see it implemented in react-query is to leverage its cache and request deduplication. If I query multiple endpoints in my fetch function, I may end up issuing multiple identical requests, because the library cannot deduplicate them.

What do you think?

**Reply by tannerlinsley** [reply]
For now, you'll need to either:
- use a single useQuery and Promise.all to resolve multiple sets of data and cache them together, or
- use separate components for each piece of data, each with their own useQuery

In the future, we may add support for caching multiple queries in a single hook that is compliant with the rules of hooks. Until then, these are your best options.
